activerecord/lib/active_record/autosave_association.rb

```diff
@@ -154,8 +154,12 @@ def self.valid_options
 
     module ClassMethods # :nodoc:
       private
-        def define_non_cyclic_method(name, &block)
-          return if method_defined?(name, false)
+        def define_non_cyclic_method(name, reflection, &block)
+          @_method_defined_reflections ||= {}
+          @_method_defined_reflections[reflection] ||= {}
+          return if @_method_defined_reflections[reflection][name]
+
+          @_method_defined_reflections[reflection][name] = true
```

Comment:
I'm open to feedback on optimising this if performance is of concern here. Essentially we're now defining by reflection instead of simply checking if the method has been defined prior. In a majority of cases it'll be defined once.